pkg/api/validate/openshiftcluster_validatedynamic.go

```diff
@@ -64,42 +95,34 @@ func (*azureClaim) Valid() error {
 	return fmt.Errorf("unimplemented")
 }
 
-type openShiftClusterDynamicValidator struct {
-	log *logrus.Entry
-	env env.Interface
-
-	oc *api.OpenShiftCluster
+func (dv *openShiftClusterDynamicValidator) Setup(ctx context.Context) error {
+	r, err := azure.ParseResourceID(dv.oc.ID)
+	if err != nil {
+		return err
+	}
 
-	fpAuthorizer refreshable.Authorizer
+	dv.spAuthorizer, err = dv.ValidateServicePrincipalProfile(ctx)
+	if err != nil {
+		return err
+	}
 
-	fpPermissions     authorization.PermissionsClient
-	spPermissions     authorization.PermissionsClient
-	spProviders       features.ProvidersClient
-	spUsage           compute.UsageClient
-	spVirtualNetworks network.VirtualNetworksClient
+	dv.spPermissions = authorization.NewPermissionsClient(r.SubscriptionID, dv.spAuthorizer)
+	dv.spProviders = features.NewProvidersClient(r.SubscriptionID, dv.spAuthorizer)
+	dv.spUsage = compute.NewUsageClient(r.SubscriptionID, dv.spAuthorizer)
+	dv.spVirtualNetworks = network.NewVirtualNetworksClient(r.SubscriptionID, dv.spAuthorizer)
+	return nil
 }
 
 // Dynamic validates an OpenShift cluster
 func (dv *openShiftClusterDynamicValidator) Dynamic(ctx context.Context) error {
 	// TODO: Dynamic() should work on an enriched oc (in update), and it
 	// currently doesn't.  One sticking point is handling subnet overlap
 	// calculations.
-
-	r, err := azure.ParseResourceID(dv.oc.ID)
+	err := dv.Setup(ctx)
```

Comment:
monitor, or possibly the diagnostic GA, will not run all the checks from within the same function as it is now in `Dynamic`. Therefore the setup part also needs to be taken out as a separate method